src/calculus/sum.jl

```diff
@@ -0,0 +1,60 @@
+export Sum
+
+"""
+**Sum of functions**
+
+    Sum(f₁,…,fₖ)
+
+Given functions `f₁` to `fₖ`, returns their sum
+
+```math
+g(x) = ∑_\{i=1\}^k f_i(x).
+```
+
+The functions `fᵢ` should all have the same domain, which will be the domain of
+the result. If all `fᵢ` have gradients then this function will also have a
+gradient.
+
+"""
+
+immutable Sum{T <: Tuple} <: ProximableFunction fs::T end
+
+Sum(fs::Vararg{ProximableFunction}) = Sum((fs...))
+
+# note: is_prox_accurate false because prox in general doesn't exist?
+is_prox_accurate(f::Sum) = false
+is_convex(f::Sum) = all(is_convex.(f.fs))
+is_set(f::Sum) = all(is_set.(f.fs))
+is_singleton(f::Sum) = all(is_singleton.(f.fs))
+is_cone(f::Sum) = all(is_cone.(f.fs))
+is_affine(f::Sum) = all(is_affine.(f.fs))
+is_smooth(f::Sum) = all(is_smooth.(f.fs))
+is_quadratic(f::Sum) = all(is_quadratic.(f.fs))
+is_generalized_quadratic(f::Sum) = all(is_generalized_quadratic.(f.fs))
+is_strongly_convex(f::Sum) = all(is_strongly_convex.(f.fs))
+
+function (sumobj::Sum)(x)
+    sum = 0.0
+    for f in sumobj.fs
+        sum += f(x)
+    end
+    sum
+end
+
+function gradient!{T}(grad::T, sumobj::Sum, x::T)
+    # gradient of sum is sum of gradients
+    val = 0.0
+    # to keep track of this sum, i may not be able to
+    # avoid allocating an array
+    temp = zeros(grad)
+    for f in sumobj.fs
+        val += gradient!(temp, f, x)
+        grad[:] += temp
```

Comment:
Maybe here `grad .+= temp` is more efficient: this `[:]` notation to make in-place operations may be do unnecessary allocations, you can check that using @time or BenchmarkTools.

Reply:
it seems like they are equivalent, but i'd rather lean on the dot notation just because it looks nicer.